### PR TITLE
Make @cached_method support ordinal args

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

Currently @cached_method has the limitation that one can only use kwargs. This is an artificial limitation. Additionally one only discovers this at runtime, rather than definition time, which is problematic.

This is a bit annoying as a user. More importantly, if one was in the situation were you wanted to apply @cached_method to something in the codebase to solve a performance issue, you would also need to go and update all callsites to use kwargs, which is thrashy.

This PR canonicalizes args and kwargs into a single kwargs dictionary for the purposes of caching values. 

## How I Tested These Changes

Included unit tests and BK

### CODE REVIEW

**Comment by schrockn** [comment]
* **#20212** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20212?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @schrockn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py [hunk 1]
```diff
@@ -96,3 +91,35 @@ def stuff(self, a=None, b=None):
...
+    b1 = obj.stuff(None, 1)
+    a2 = obj.stuff(2, None)
+    b2 = obj.stuff(None, 2)
+
+    assert a1 != b1
+    assert a1 != a2
+    assert b1 != b2
+
+
+def test_doc_string_scenario_canonicalized_cache_entry() -> None:
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py:113]: I'm having trouble relating this test name to what's being tested.

File: python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py [hunk 2]
```diff
@@ -96,3 +91,35 @@ def stuff(self, a=None, b=None):
...
+        def a_method(self, arg1: str, arg2: int) -> str:
+            return arg1 + str(arg2)
+
+    obj = MyClass()
+    assert obj.a_method(arg1="a", arg2=5) == "a5"
+    assert obj.a_method(arg2=5, arg1="a") == "a5"
+    assert obj.a_method("a", 5) == "a5"
+
+    # only one entry
+    assert len(obj.__dict__) == 1
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py:125]: Should this be
```
    assert len(obj.__dict__["a_method" + CACHED_METHOD_FIELD_SUFFIX]) == 1
```

File: python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py [hunk 3]
```diff
@@ -96,3 +91,35 @@ def stuff(self, a=None, b=None):
...
+
+    obj = MyClass()
+    a1 = obj.stuff(1, None)
+    b1 = obj.stuff(None, 1)
+    a2 = obj.stuff(2, None)
+    b2 = obj.stuff(None, 2)
+
+    assert a1 != b1
+    assert a1 != a2
+    assert b1 != b2
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py:110]: Maybe this is covered by the test below, but I would have expected to see an assert like
```python
assert obj.stuff(1, None) is obj.stuff(a=1, b=None)
```

**Review by sryza — COMMENTED** [review]
Agree that this would be a great user experience improvement for this method.

My recollection of why I omitted this in the first place:
- I believe `functools.lru_cache` does not do this canonicalization
- The [implementation](https://github.com/python/cpython/blob/main/Lib/functools.py#L491) of `functools.lru_cache` seems very performance-optimized, so my guess was that they didn't do the canonicalization because it would slow things down

Which is to say that I think we should consider either benchmarking the change or documenting that this might be slower than `functools.lru_cache`.

**Review by alangenfeld — CHANGES_REQUESTED** [review]
I think this is fine even if its less performant when args are used. Just want to know by how much and try to ensure comments are put in the right place so if the args related stuff shows up in a profiler its a quick resolution to knowing switching to kwargs will speed things up. 

previous optimization PRs
https://github.com/dagster-io/dagster/pull/13229
https://github.com/dagster-io/dagster/pull/13651

I unfortunately didn't write down exactly how i did it, but I am pretty sure it was using 
`timeit` https://docs.python.org/3/library/timeit.html with
* defining the cached_method in `setup` 
* doing M timeit iterations of of  N calls of the fn

**Comment by schrockn** [comment]
The following script:

```python
import timeit

class OnlyKwargs:
    def add(self, a: int, b: int) -> int:
        return a + b

class MixedArgs:
    def add(self, a: int, b: int) -> int:
        return a + b

class OnlyOrdinalArgs:
    def add(self, a: int, b: int) -> int:
        return a + b


N = 100_000_000

only_kwargs_t = timeit.Timer('obj.add(a=1, b=2)', setup='obj = OnlyKwargs()', globals=globals())
print(f"Only kwargs: {only_kwargs_t.timeit(N)}") 

mixed_args_t = timeit.Timer('obj.add(1, b=2)', setup='obj = MixedArgs()', globals=globals())
print(f"Mixed kwargs: {mixed_args_t.timeit(N)}") 

ordinal_args_t = timeit.Timer('obj.add(1, 2)', setup='obj = OnlyOrdinalArgs()', globals=globals())
print(f"Only ordinals: {ordinal_args_t.timeit(N)}") 
```


Against these changes:

```
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs: 3.4493765420047566
Mixed kwargs: 3.1732202499406412
ordinal kwargs: 2.4479680419899523
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs: 3.459478833945468
Mixed kwargs: 3.176478792098351
Only ordinals: 2.4514362500049174
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs: 3.451715874951333
Mixed kwargs: 3.1782272909767926
Only ordinals: 2.4466074580559507
```

Against https://github.com/dagster-io/dagster/pull/20213

```
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-inline-fasttypes-set) ✗ python time_cached_method.py
Only kwargs: 3.4764964579371735
Mixed kwargs: 3.3277986249886453
Only ordinals: 2.428189832950011
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-inline-fasttypes-set) ✗ python time_cached_method.py
Only kwargs: 3.486207833979279
Mixed kwargs: 3.1910982499830425
Only ordinals: 2.483242917107418
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-inline-fasttypes-set) ✗ python time_cached_method.py
Only kwargs: 3.4733797089429572
Mixed kwargs: 3.198128000018187
Only ordinals: 2.4783102920046076
```


Against master (`92cca6c9924112004e83d8f9d37e25e7e28c2f27`)
```
(dagster-3.11.5-2024-02-25) ➜  dagster git:(master) ✗ python time_cached_method.py
Only kwargs: 3.4456689580110833
(dagster-3.11.5-2024-02-25) ➜  dagster git:(master) ✗ python time_cached_method.py
Only kwargs: 3.4511710830265656
(dagster-3.11.5-2024-02-25) ➜  dagster git:(master) ✗ python time_cached_method.py
Only kwargs: 3.4581523339729756
```

Only ordinals substantially faster versus kwargs. 

This is because ython internals are faster with ordinals and then dominate any overhead we add. This is replicated by other people:

https://stackoverflow.com/questions/67478584/performance-difference-between-positional-and-keyword-arguments-in-python

File: python_modules/dagster/dagster/_utils/cached_method.py [hunk 4]
```diff
@@ -48,11 +48,15 @@ def a_method(self, arg1, arg2):
...
             obj.a_method("a", 5)
 
-    With this decorator, the first two would point to the same cache entry, and non-kwarg arguments
-    are not allowed.
+    With this decorator, keyword and non-keyword arg usage is canonicalized and the above
+    calls result in the same cache entry
     """
     cache_attr_name = method.__name__ + CACHED_METHOD_FIELD_SUFFIX
 
+    arg_names = get_arg_names(method)
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_utils/cached_method.py]: I moved this to be created on demand 👍🏻

File: python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster_tests/general_tests/utils_tests/test_cached_method.py:110]: Great call thx

**Comment by alangenfeld** [comment]
Looks like you missed using `@cached_method` in your profiling script.

I also think we want to call the function a number of times under test to evaluate the cost of the cache lookup on hit past the initial fill invocation. 

With 
```
import timeit
from dagster._utils.cached_method import cached_method


class TestObj:
    @cached_method
    def add(self, a: int, b: int) -> int:
        return a + b


N = 1_000
M = 1_000

only_kwargs_t = timeit.Timer(
    "[obj.add(a=1, b=2) for _ in range(M)]",
    setup="obj = TestObj()",
    globals=globals(),
)
print(f"Only kwargs: {only_kwargs_t.timeit(N)}")

mixed_args_t = timeit.Timer(
    "[obj.add(1, b=2) for _ in range(M)]",
    setup="obj = TestObj()",
    globals=globals(),
)
print(f"Mixed kwargs: {mixed_args_t.timeit(N)}")

ordinal_args_t = timeit.Timer(
    "[obj.add(1, 2) for _ in range(M)]",
    setup="obj = TestObj()",
    globals=globals(),
)
print(f"Only ordinals: {ordinal_args_t.timeit(N)}")
```

On this PR I get 
```
Only kwargs: 0.698514708026778
Mixed kwargs: 0.878142207977362
Only ordinals: 0.8298114589997567
```

**Comment by schrockn** [comment]
Results in this branch: 

```python
import timeit

from dagster._utils.cached_method import cached_method

class OnlyKwargs:
    def add(self, a: int, b: int) -> int:
        return a + b

class MixedArgs:
    def add(self, a: int, b: int) -> int:
        return a + b

class OnlyOrdinalArgs:
    def add(self, a: int, b: int) -> int:
        return a + b

class OnlyKwargsCached:
    @cached_method
    def add(self, a: int, b: int) -> int:
        return a + b

class MixedArgsCached:
    @cached_method
    def add(self, a: int, b: int) -> int:
        return a + b

class OnlyOrdinalArgsCached:
    @cached_method
    def add(self, a: int, b: int) -> int:
        return a + b


N = 10_000_000

only_kwargs_t = timeit.Timer('obj.add(a=1, b=2)', setup='obj = OnlyKwargs()', globals=globals())
print(f"Only kwargs uncached: {only_kwargs_t.timeit(N)}") 

mixed_args_t = timeit.Timer('obj.add(1, b=2)', setup='obj = MixedArgs()', globals=globals())
print(f"Mixed kwargs uncached: {mixed_args_t.timeit(N)}") 

ordinal_args_t = timeit.Timer('obj.add(1, 2)', setup='obj = OnlyOrdinalArgs()', globals=globals())
print(f"Only ordinals uncached: {ordinal_args_t.timeit(N)}") 

only_kwargs_cached_t = timeit.Timer('obj.add(a=1, b=2)', setup='obj = OnlyKwargsCached()', globals=globals())
print(f"Only kwargs cached: {only_kwargs_cached_t.timeit(N)}") 

mixed_args_cached_t = timeit.Timer('obj.add(1, b=2)', setup='obj = MixedArgsCached()', globals=globals())
print(f"Mixed kwargs cached: {mixed_args_cached_t.timeit(N)}") 

ordinal_args_cached_t = timeit.Timer('obj.add(1, 2)', setup='obj = OnlyOrdinalArgsCached()', globals=globals())
print(f"Only ordinals cached: {ordinal_args_cached_t.timeit(N)}") 
```

```
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs uncached: 0.3528227499919012
Mixed kwargs uncached: 0.32212733395863324
Only ordinals uncached: 0.24858516606036574
Only kwargs cached: 6.905509624979459
Mixed kwargs cached: 8.312129166908562
Only ordinals cached: 7.952683624927886
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs uncached: 0.36741120798978955
Mixed kwargs uncached: 0.340193499927409
Only ordinals uncached: 0.25531091704033315
Only kwargs cached: 6.7889418749837205
Mixed kwargs cached: 8.676265459042042
Only ordinals cached: 8.03488574991934
(dagster-3.11.5-2024-02-25) ➜  dagster git:(cached-method-support-args) ✗ python time_cached_method.py
Only kwargs uncached: 0.36696395894978195
Mixed kwargs uncached: 0.34307924995664507
Only ordinals uncached: 0.26282441697549075
Only kwargs cached: 6.846036625094712
Mixed kwargs cached: 8.666449791984633
```

**Comment by schrockn** [comment]
So @alangenfeld call it 15% overhead. Seems ok to me.

File: python_modules/dagster/dagster/_utils/cached_method.py [hunk 5]
```diff
@@ -61,7 +63,29 @@ def _cached_method_wrapper(self: S, *args: P.args, **kwargs: P.kwargs) -> T:
...
+            translated_kwargs = {}
+            for arg_ordinal, arg_value in enumerate(args):
+                arg_name = arg_names[arg_ordinal]
+                translated_kwargs[arg_name] = arg_value
+            if kwargs:
+                # only copy if both args and kwargs were passed
+                canonical_kwargs = {**translated_kwargs, **kwargs}
+            else:
+                # no copy
+                canonical_kwargs = translated_kwargs
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_utils/cached_method.py:83]: [1]

**Comment by schrockn** [comment]
### Merge activity

* **Mar 6, 2:28 PM EST**: @schrockn started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/20212).

**Pull request merged**